Make NextBus coordinator more resilient and efficient

ViViDboarder opened the pull request:
## Proposed change
Resolves issues where one request failing will prevent all agency predictions to fail. This also removes redundant requests for predictions that share the same stop.

This involves updating the dependency to the latest version via #126230 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #123625
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ViViDboarder commented:
Failure is unrelated to this integration and appears to come from dev. I can rebase when that’s fixed.

joostlek commented:
Let's bump the dependency in a separate PR

joostlek commented on homeassistant/components/nextbus/__init__.py:
```diff
@@ -13,15 +13,19 @@
-        entry_agency
+    coordinator: NextBusDataUpdateCoordinator | None = hass.data.setdefault(
+        DOMAIN, {}
+    ).get(
+        coordinator_key,
     )
     if coordinator is None:
         coordinator = NextBusDataUpdateCoordinator(hass, entry_agency)
-        hass.data[DOMAIN][entry_agency] = coordinator
+        hass.data[DOMAIN][coordinator_key] = coordinator
 
-    coordinator.add_stop_route(entry.data[CONF_STOP], entry.data[CONF_ROUTE])
+    coordinator.add_stop_route(entry_stop, entry.data[CONF_ROUTE])
```
Can we maybe add tests for something I want to verify? Once you create a coordinator in a config entry context, it will bind the coordinator to the config entry and it will shutdown the coordinator when the config entry unloads 

https://github.com/home-assistant/core/blob/f006716173eb16848b0bcf53d8309fcdea803019/homeassistant/helpers/update_coordinator.py#L123-L124

So I think this setup might not even work here and cause the other config entries to stop working when for example a config entry is deleted

ViViDboarder replied:
Sorry. I'm not following.

On setup the coordinator is being added to the domain under a key that is possibly shared between different entities.

In the remove config function, the relevant coordinator is looked up, the entity specific data is popped out, then there is a check to see if there are remaining routes the coordinator is still responsible for, only if this is empty is it deleted.

What would a test like this look like?

joostlek replied:
But check the code that I linked, the config entry that created the coordinator would shut down the coordinator while the entry got unloaded.

Set up 2 entries which share that coordinator, then unload one, skip time and check if the coordinator still works like expected

ViViDboarder replied:
Oh, hmm. I've been using this pattern since the original creation of this component. I figured this was `async_unload_entry` was supposed to handle, so I managed the logic of when to stop listening there. It looks like this `async_on_unload` was added only last year and may have introduced a bug.

It looks like when implemented it was suggested to make it "opt in" to auto cleanup, but instead no optionally was provided, with the intent of making sure this "just works" for typical use cases. This could be resolved by adding an "opt out" mechanism which would preserve the "just works" aspect of "fixing" zombie coordinators.

I've managed another way to do it by overriding the unload function and checking there.

I've got a test ready but trying to figure out how to "skip time". Any pointers?

ViViDboarder replied:
Figured it out. I was about to confirm the issue you mentioned with a test and then verify my patch resolves it.

joostlek replied:
Use the freezer! Im on mobile so I can't point to code right now. But a quick search will yield results.

I think it would be nice if we can go for a separate setup (so every config entry it's own coordinator). This way we know everything works like expected and then we make sure we have tests. And then we can always discover ways to optimise this. There are other classes (like the Base coordinator protocol) which don't have that limitation, so maybe that works for you. But I think it would be nice to get it to a working state before we start to optimise

ViViDboarder replied:
Well, the efficiency and the working state fix are one in the same here. There is an undocumented API that is being used and it appears that too many requests and it starts to 404. This patch reduces reuse of the coordinator where it doesn't add efficiency as well as allows reuse when it will. This should reduce 404s, but also prevent a 404 from killing updates from unrelated entities.

joostlek replied:
Haha okay, I was scared when you said "I've managed another way to do it by overriding the unload function and checking there.". I thought you did some crazy thing to avoid the async_on_unload, hence my comment. 

I'll review later

ViViDboarder replied:
Oh. Not too crazy. It’s a public method, so I added a check and then call super.

ViViDboarder committed:
Make NextBus coordinator more resilient and efficient

Resolves issues where one request failing will prevent all agency
predictions to fail. This also removes redundant requests for
predictions that share the same stop.

ViViDboarder committed:
Add unload entry test

ViViDboarder committed:
Prevent shutdown if the coordinator is still needed

ViViDboarder commented on homeassistant/components/nextbus/coordinator.py:
```diff
@@ -48,27 +48,63 @@ def has_routes(self) -> bool:
 
-        _route_stops = set(self._route_stops)
-        self.logger.debug("Updating data from API. Routes: %s", str(_route_stops))
+        _stops_to_route_stops: dict[str, set[RouteStop]] = {}
+        for route_stop in self._route_stops:
+            _stops_to_route_stops.setdefault(route_stop.stop_id, set()).add(route_stop)
```
For solutions I only ever need to create once, I tend to use `setdefault` rather than `defaultdict`, mostly just because I find it simpler. Since I’m not defaulting anywhere else, I’d prefer to keep it this way. 

Do you have strong feelings about changing it?

joostlek replied:
My main reason would be to remove the `.setdefault` and just directly do `.add` which looks cleaner imo, but if you're not up for it I am not going to block it
